test/Test/MockApi.hs

```diff
@@ -0,0 +1,233 @@
+      let unsatisfiedSingleExpectations = filter isSingular remainingExpectations
+
+      if null unsatisfiedSingleExpectations
+        then pure ctx
+        else
+          sendIO . assertFailure $
+            "Test completed with unsatisfied expectations: \n  "
+              <> intercalate "\n  " (map (\(ApiExpectation _ req _) -> show req) unsatisfiedSingleExpectations)
+    R other -> alg (runMockApiC . hdl) (R other) ctx
+
+isSingular :: ApiExpectation -> Bool
+isSingular (ApiExpectation Once _ _) = True
+isSingular (ApiExpectation Always _ _) = False
+
+-- | Matches a request to an expectation.  This function basically exists to
+-- extract and compare the runtime constructor of the two arguments so that
+-- arbitrary ones can be compared even if the types wouldn't match.
+matchExpectation :: FossaApiClientF a -> ApiExpectation -> Maybe (ApiResult a)
+matchExpectation a@(GetApiOpts) (ApiExpectation _ b@(GetApiOpts) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetIssues{}) (ApiExpectation _ b@(GetIssues{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetLatestBuild{}) (ApiExpectation _ b@(GetLatestBuild{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetLatestScan{}) (ApiExpectation _ b@(GetLatestScan{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetOrganization) (ApiExpectation _ b@(GetOrganization) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetProject{}) (ApiExpectation _ b@(GetProject{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(GetScan{}) (ApiExpectation _ b@(GetScan{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(UploadAnalysis{}) (ApiExpectation _ b@(UploadAnalysis{}) resp) = resp <$ guard (a == b)
+matchExpectation a@(UploadContributors{}) (ApiExpectation _ b@(UploadContributors{}) resp) = resp <$ guard (a == b)
+matchExpectation _ _ = Nothing
+
+-- | Handles a request in the context of the mock API.
+handleRequest ::
+  ( Has (State [ApiExpectation]) sig m
+  ) =>
+  FossaApiClientF a ->
+  m (Maybe (ApiResult a))
+handleRequest req = do
+  expectations <- get
+  case testExpectations expectations of
+    Just (resp, expectations') -> do
+      put expectations'
+      pure (Just resp)
+    Nothing ->
+      pure Nothing
+  where
+    testExpectations [] = Nothing
+    testExpectations (expectation : rest) =
+      case matchExpectation req expectation of
+        Nothing -> fmap (expectation :) <$> testExpectations rest
+        Just resp ->
+          if isSingular expectation
+            then Just (resp, rest)
+            else Just (resp, expectation : rest)
+
+type FossaApiClientMockC = SimpleC FossaApiClientF
+
+-- | Run an action with a mock API client that tracks call expectations
+runApiWithMock ::
+  ( Has (Lift IO) sig m
+  , Has Diagnostics sig m
+  , Has MockApi sig m
+  ) =>
+  FossaApiClientMockC m a ->
+  m a
+runApiWithMock f = do
+  result <- interpret runRequest f
+  assertAllSatisfied
+  pure result
+  where
+    runRequest req = do
```

Comment:
[Optional] [Minor] Let's add types for runRequest function in this where bock, so it's easier to read